source/extensions/quic_listeners/quiche/active_quic_listener.cc

```diff
@@ -112,14 +117,37 @@ void ActiveQuicListener::onData(Network::UdpRecvData& data) {
                                   /*packet_headers=*/nullptr, /*headers_length=*/0,
                                   /*owns_header_buffer*/ false);
   quic_dispatcher_->ProcessPacket(self_address, peer_address, packet);
+
+  if (!schedule_process_buffered_enabled_) {
+    // On event loop runs where this listener reads or has packets posted to it, process buffered
+    // packets on the next loop. This ensures that the limit of processed client hellos per loop is
+    // honored, and that buffered client hellos are processed in a timely fashion.
+    schedule_process_buffered_->scheduleCallbackNextIteration();
+    schedule_process_buffered_enabled_ = true;
+  }
 }
 
-void ActiveQuicListener::onReadReady() {
+void ActiveQuicListener::onReadReady() {}
+
+void ActiveQuicListener::processBuffered() {
+  schedule_process_buffered_enabled_ = false;
+
   if (!enabled_.enabled()) {
     ENVOY_LOG(trace, "Quic listener {}: runtime disabled", config_->name());
     return;
   }
+
+  if (quic_dispatcher_->HasChlosBuffered()) {
+    event_loop_chlo_buffered_++;
+  }
+
   quic_dispatcher_->ProcessBufferedChlos(kNumSessionsToCreatePerLoop);
+
+  // If there were more buffered than the limit, schedule again for the next event loop.
+  if (quic_dispatcher_->HasChlosBuffered()) {
+    schedule_process_buffered_->scheduleCallbackNextIteration();
```

Comment:
I think you need if (!schedule_process_buffered_->enabled()) here since otherwise you can end up starving this callback, but I'm not 100% sure.  I think this scenario ends up being similar to the timer2 in TimerImplTest.TimerOrderAndReschedule; the rescheduling of timer2 which was in the active list for that iteration results in timer2 running on a future iteration of the event loop.

We may want to change SchedulableCallback::scheduleCallbackNextIteration to be a no-op in cases where the callback is already enabled.

Reply:
https://github.com/envoyproxy/envoy/pull/13097